lib/kcerts/ssh.go

```diff
@@ -0,0 +1,197 @@
+package kcerts
+
+import (
+	"crypto/rand"
+	"crypto/rsa"
+	"encoding/json"
+	"fmt"
+	"github.com/enfabrica/enkit/lib/cache"
+	"github.com/mitchellh/go-homedir"
+	"golang.org/x/crypto/ssh"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"regexp"
+	"strconv"
+	"strings"
+	"time"
+)
+
+const CAPrefix = "@cert-authority"
+const SSHDir = ".ssh"
+const KnownHosts = "known_hosts"
+
+const SSHCacheKey = "enkit_ssh_cache_key"
+
+// FindSSHDir will find the users ssh directory based on $HOME. If $HOME/.ssh does not exist
+// it will attempt to create it.
+func FindSSHDir() (string, error) {
+	hDir, err := homedir.Dir()
+	if err != nil {
+		return "", fmt.Errorf("could not find the home directory: %w", err)
+	}
+	sshDir := filepath.Join(hDir, SSHDir)
+	if err := os.Mkdir(sshDir, 0700); err != nil && !os.IsExist(err) {
+		return "", fmt.Errorf("could not create directory %s: %w", sshDir, err)
+	}
+	return sshDir, nil
+}
+
+// AddSSHCAToClient adds a public key to the $HOME/.ssh/known_hosts in the ssh-cert x509.1 format.
+// For each entry, it adds an additional line and does not concatenate
+func AddSSHCAToClient(publicKey ssh.PublicKey, hosts []string, sshDir string) error {
+	caPublic := string(ssh.MarshalAuthorizedKey(publicKey))
+	knownHosts := filepath.Join(sshDir, KnownHosts)
+	knownHostsFile, err := os.OpenFile(knownHosts, os.O_RDWR|os.O_CREATE|os.O_APPEND, 0644)
+	defer knownHostsFile.Close()
+	if err != nil {
+		if os.IsNotExist(err) {
+			return fmt.Errorf("could not create known_hosts file: %w", err)
+		}
+		return err
+	}
+	existingKnownHostsContent, err := ioutil.ReadAll(knownHostsFile)
+	if err != nil {
+		return fmt.Errorf("error reading %s: %w", knownHosts, err)
+	}
+	for _, dns := range hosts {
+		// caPublic terminates with a '\n', added by ssh.MarshalAuthorizedKey
+		publicFormat := fmt.Sprintf("%s %s %s", CAPrefix, dns, caPublic)
+		if strings.Contains(string(existingKnownHostsContent), publicFormat) {
+			continue
+		}
+		_, err = knownHostsFile.WriteString(publicFormat)
+		if err != nil {
+			return fmt.Errorf("could not add key %s to known_hosts file: %w", publicFormat, err)
+		}
+	}
+	return nil
+}
+
+type sshCache struct {
+	PID       int       `json:"pid"`
+	Sock      string    `json:"sock"`
+	TimeStamp time.Time `json:"time_stamp"`
+}
+
+// FindSSHAgent Will start the ssh agent in the interactive terminal if it isn't present already as an environment variable
+// Currently only outputs the env and does not persist it across terminals
+func FindSSHAgent(store cache.Store, ttl time.Duration) (string, int, error) {
```

Comment:
meta: I really like this function and the improvements you've made. I would however break this down in multiple functions. Something like this:

```
type Agent struct {
  Pid int
  Socket string
}

func FindSSHAgent(store cache.Store) (*Agent, error) {
  agent, err := LoadAgentFromEnv() // Parses env variables to find an agent.
  if err == nil && agent.IsValid() {
    return agent, nil
  }

  agent, cache, err = LoadAgentFromCache(store)
  defer store.Rollback(cache)
  if err == nil && agent.IsValid() {
    return agent, nil
  }

  agent, err = CreateAgent()
  if err != nil {
    return err
  }

  StoreAgentInCache(agent, cache)
  return nil
}
```

also: both re-using an existing agent or loading from cache (or storing in cache) are "optimizations", the code should not fail or return error if those things fail.

Reply:
By the way: if you do this, the ttl parameter is no longer needed? Basically: if you find a working agent, no matter what, you can reuse it.

Reply:
`IsValid` could be implemented as something like:
```
	conn, err := net.Dial("unix", agent.Socket)
	defer conn.Close()
        return err != nil
```

you could even have an `Open()` on the `Agent` object to later on use it.